contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+        markets.forEach(async (market: PublicKey) => {
+            if (!wsConnected) {
+                // Fetch via RPC for the openbook market
+                const _market = await openBookProgram.account.market.fetch(
+                    market
+                );
+                const sides = [
+                    {
+                        pubKey: _market.asks,
+                        side: 'asks',
+                    },
+                    {
+                        pubKey: _market.bids,
+                        side: 'bids',
+                    },
+                ];
+                // Setup Websocket subscription for the two sides
+                try {
+                    const subscriptionId = sides.map((side) => provider.connection.onAccountChange(
+                        side.pubKey,
+                        (updatedAccountInfo, ctx) => {
+                            consumeOrderBookSide(side.side, updatedAccountInfo, ctx);
+                        },
+                        'processed'
+                    )
+                    );
+                    return subscriptionId;
+                } catch (err) {
+                    setWsConnected(false);
+                }
+            }
+            // For map handling
+            return null;
+        });
+    };
+
+    const cancelOrder = useCallback(async (order: OpenOrdersAccountWithKey) => {
+        if (!proposal || !markets) return;
+        const marketAccount = isPassMarket
+            ? { publicKey: proposal.account.openbookPassMarket, account: markets.pass }
+            : { publicKey: proposal.account.openbookFailMarket, account: markets.fail };
+        const txs = await cancelOrderTransactions(
+            new BN(order.account.accountNum),
+            marketAccount,
+        );
+
+        if (!wallet.publicKey || !txs) return;
+
+        try {
+            //settle it right away
+            const settleTxs = await settleFundsTransactions(
+                order.account.accountNum,
+                isPassMarket,
+                proposal,
+                marketAccount
+            );
+
+            if (!txs) return;
+
+            const txsSent = await sender.send([...txs, ...settleTxs]);
+            if (txsSent.length !== 0) {
+                //update order in state
+                const cancelledOrderIndex = orders.findIndex(o => o.account.accountNum == order.account.accountNum);
+                orders[cancelledOrderIndex].account.openOrders[0].isFree = 1;
+                setOrders(orders);
+            }
+        } catch (err) {
+            console.error(err);
+        }
+    }, [markets, proposal]);
+
+    useEffect(() => {
+        if (!wsConnected) {
+            listenOrderBook();
+        }
+    }, [wsConnected]);
+    useEffect(() => {
+        fetchMarketsInfo();
+    }, [proposal]);
+
+    const memoValue = useMemo(() => {
+        return {
+            markets,
+            orders,
+            orderBookObject,
```

Comment:
The only other thing I'll note is that we've got other things coupled to this concept which aren't passed into this context, like cranking. I've also noted we've got some disparate pieces in hooks as well (useOpenbookTwap, useOpenbook) not sure if we should make some attempt to make sense of why or where we're placing things or even break those down into more simple creations so they're better reasoned and more accessible for people to work on?

Reply:
Yeah I wasn't sure about the cranking so I left it in the component, but I could move it here as well.